File: 043_multiply_strings/Solution2/Solution.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func multiply(num1 string, num2 string) string {
	if num1 == "0" || num2 == "0" {
		return "0"
	}
	n1, n2 := len(num1), len(num2)
	tmpRes := make([]int, n1 + n2)

	for i := n1 - 1; i >= 0; i-- {
		x := num1[i] - '0'
		for j := n2 - 1; j >= 0; j-- {
			y := num2[j] - '0'
			tmp := int(x * y) + tmpRes[i + j + 1]
			tmpRes[i + j + 1] = tmp % 10
			tmpRes[i + j] += tmp / 10
		}
	}

	var result []string
	for i, v := range tmpRes {
		if i == 0 && v == 0 {
			continue
		}
		result = append(result, strconv.Itoa(v))
	}

	return strings.Join(result, "")
}

func main() {
	fmt.Println(multiply("123", "45"))
}

File: 043_multiply_strings/Solution2/Solution.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        if num1 == '0' or num2 == '0': return '0'
        n1, n2 = len(num1), len(num2)
        result = [0] * (n1 + n2)

        for i in range(n1 - 1, -1, -1):
            x = int(num1[i])
            for j in range(n2 - 1, -1, -1):
                y = int(num2[j])
                tmp = x * y + result[i + j + 1]
                result[i + j + 1] = tmp % 10
                result[i + j] += tmp // 10

        if result[0] == 0: result = result[1:]
        return ''.join([str(i) for i in result])


if __name__ == '__main__':
    solution = Solution()
    print(solution.multiply("123", "45"))
